first draft of haskell parser (#2)
jnduli/blog_jnduli.co.ke: content/drafts/first_parser_in_haskell.rst

############################################
First Haskell Parser (for Kindle Highlights)
############################################

:date: 2019-04-27 19:00
:tags: linux, product, review
:category: Computer
:slug: huion-420-driver-setup
:author: John Nduli
:status: draft

Kindle saves highlights, bookmarks and notes in a `My Clippings.txt`

> file. I wanted a way to parse thing and filter out the content I wanted.

Reviewer: ```suggestion
file. I wanted a way to parse it and filter out the content I wanted.
```

> I also wanted to work on a parser in haskell, so this is the result of
> that.

Reviewer: ```suggestion
I also wanted to work on a parser in haskell, so this was a great
opportunity to get my hands dirty.
```

> I used stack to manage the project by running the commands:

Reviewer: ```suggestion
I used stack to manage the project by running:
```

The `exampleString` above contains two different highlights each
separated with `==========`. To get a single group, I'd look for this
separator and collect the groups before this. To do this in parsec:

.. code-block:: haskell

    import Text.Parsec

    eogString :: String
    eogString = "=========="

    endOfGroup :: Parsec String st String
    endOfGroup = string eogString

    kindleGroup :: Parsec String st String
    kindleGroup = manyTill anyChar (try endOfGroup)

    test = parse kindleGroup "failed" exampleString

Running `test` in ghci returns `Right "this\nis\ngood\n"`.

`endOfGroup` is a parser that only matches the `eogString`. This is
combined with the `kindleGroup` to get all characters (using `anyChar`)
until the `endOfGroup` is found.

All the groups can be got by building on this through:

.. code-block:: haskell

    groups :: Parsec String st [String]
    groups = do
      first <- kindleGroup
      next  <- remainingGroups
      return (first : next)

    remainingGroups :: Parsec String st [String]
    remainingGroups = (char '\n' >> groups) <|> return []

    test = parse groups "fail" (exampleString)

> which not results in `Right ["this\nis\ngood\n","anther\ngroup\n"]`.

Reviewer: ```suggestion
which results in `Right ["this\nis\ngood\n","anther\ngroup\n"]`.
```

Groups gets a group from the string, and passes the rest of the string

> to the remainingGroups function. This inturn checks if the first

Reviewer: ```suggestion
to the remainingGroups function. This in turn checks if the first
```

        char '-'
        end <- many1 digit
        return [start, end]

    highlight :: Parsec String st [String]
    highlight = do
        t <- title
        l <- location
        title
        title
        h <- title
        let x = [t, h] ++ l
        return x

    exampleGroup = "Axiomatic (Greg Egan)\n" ++ 
        "- Your Highlight at location 3722-3722 | Added on Sunday, 28 October 2018 08:42:11\n" ++
        "\n" ++
        "mind; maybe some dreams take shape only in the\n"

This results in `Right ["Axiomatic (Greg Egan)","mind; maybe some dreams take shape only in the","3722","3722"]`

To parse multiple groups, we just:

.. code-block:: lua

    groups :: Parsec String st [[String]]
    groups = do
      first <- highlight
      next  <- remainingGroups
      return (first : next)

    remainingGroups :: Parsec String st [[String]]
    remainingGroups = (char '\n' >> groups) <|> return []

    test = parse groups "fail" (exampleGroup ++ "\n" ++ exampleGroup)

This returns: 

`Right [["Axiomatic (Greg Egan)","mind; maybe some dreams take shape only in the","3722","3722"],["Axiomatic (Greg Egan)","mind; maybe some dreams take shape only in the","3722","3722"]]`

> The kindle format has more nuts than what the above demonstrates, but I
> had a lot of fun working with this.

Reviewer: ```suggestion
The kindle format has more nuts than what the above demonstrates, but
nevertheless I had a lot of fun working with this.
```
